src/types.py

```diff
@@ -70,32 +70,27 @@ def unify(self, other: "Type") -> bool:
         if sign != Type.ANY_SIGN:
             assert kind == Type.K_INT
         if x.ptr_to is not None and y.ptr_to is not None:
-            if isinstance(x.ptr_to, Type) and isinstance(y.ptr_to, Type):
-                if not x.ptr_to.unify(y.ptr_to):
-                    return False
-            elif not isinstance(x.ptr_to, Type) and not isinstance(y.ptr_to, Type):
-                assert x.typemap == y.typemap and x.typemap
-                # TODO: deep resolve_typedefs (needs a typemap)
-                if not equal_types(x.ptr_to, y.ptr_to):
+            if not isinstance(x.ptr_to, Type) and not isinstance(y.ptr_to, Type):
+                assert x.typemap is not None
+                assert y.typemap is not None
+                assert x.typemap == y.typemap
+                x_ctype = resolve_typedefs(x.ptr_to, x.typemap)
+                y_ctype = resolve_typedefs(y.ptr_to, y.typemap)
+                if not equal_types(x_ctype, y_ctype):
                     return False
             else:
-                # TODO: unify Type and CType (needs a typemap)
-                # Until we get that, let's handle some easy cases though:
-                if isinstance(x.ptr_to, Type) and not isinstance(y.ptr_to, Type):
-                    type = x.ptr_to
-                    ctype = y.ptr_to
-                    typemap = y.typemap
-                elif isinstance(y.ptr_to, Type) and not isinstance(x.ptr_to, Type):
-                    type = y.ptr_to
-                    ctype = x.ptr_to
-                    typemap = x.typemap
+                if isinstance(x.ptr_to, Type):
+                    x_type = x.ptr_to
+                else:
+                    assert x.typemap is not None
+                    x_type = type_from_ctype(x.ptr_to, x.typemap)
```

Comment:
`type_from_ctype` is lossy, because there's no Type corresponding to a C struct for instance (perhaps a design mistake, but it *is* convenient to keep up the pretense that all local variables/expressions/etc. are register sized). I think maybe if `type_from_ctype(x.ptr_to, x.typemap)` returns Type.any() and `y.ptr_to` is not Type.any() we should consider returning False here?

Reply:
I think trying to unifying them is still correct?

We only use `type_from_ctype` when one of `x.ptr_to`/`y.ptr_to` is a `CType`, and when one is a `Type`. I think the only lossy case would be when *both* are `CType`s?

Let's say `x.ptr_to` is the `CType` & `y.ptr_to` is the `Type`. We lose information if `type_from_ctype(x.ptr_to, ...) == Type.any()`, and that means `x` is either a `void *` or a struct/union pointer.

If `y.ptr_to` is `Type.any()`, then that means `y` is a `void *`, so it's valid to unify it with the struct pointer. On the other hand, if `y.ptr_to` is an `float` or something, it won't be able to unify with `x.ptr_to`.

Reply:
The case I'm worried about is if x.ptr_to is a Struct and y.ptr_to is say a Type.f32(), then unification will currently succeed which seems sketchy. Claiming `void*` = `float*` also feels suspect, being non-transitive.

If x.ptr_to is Type.any() a reasonable unification would be to set x.ptr_to = y.ptr_to, no need to do type_from_ctype. Otherwise I'm dubious about the struct/void cases. But it's 3 AM so I may be missing something :)

Reply:
I think I see the issue, and I refactored this to avoid unifying with `Type.any()`'s.

However, I think in both the original code and my PR, the unify implementation has `void * == float *`. 

It seems like `void *` could be represented two ways: `Type.ptr(None)` or `Type.ptr(Type.any())`. 

In the first case, `x.ptr_to` would just be `None` so this whole `if` block I'm modifying wouldn't be hit, and the types would just unify taking `y.ptr_to`.
In the second case, with `x.ptr_to = Type.any()`, that type could unify with `y.ptr_to = Type.f32()`.